docs/education/dlt-advanced-course/lesson_1_custom_sources_restapi_source_and_restclient.py

```diff
@@ -0,0 +1,1223 @@
+    _news_source = rest_api_source(_news_config)
+    pipeline_3 = dlt.pipeline(pipeline_name='news_pipeline', destination='duckdb', dataset_name='news')
+    pipeline_3.run(_news_source)
+    print(pipeline_3.last_trace)
+    return (pipeline_3,)
+
+
+@app.cell
+def _(pipeline_3):
+    pipeline_3.dataset().news_articles.df().head()
+    return
+
+
+@app.cell(hide_code=True)
+def _(mo):
+    mo.md(r"""
+    ## **4. Add order, filtering via params**
+    We can filter articles using query parameters supported by NewsAPI:
+
+    ```python
+    "params": {
+        "q": "python",
+        "language": "en",
+        "pageSize": 20,
+    },
+    ```
+
+    - `q`: search keyword (e.g. "python")
+    - `language`: filter by article language
+    - `pageSize`: number of articles per page (max 100)
+    """)
+    return
+
+
+@app.cell
+def _(dlt, os, rest_api_source):
+    api_key_4 = os.getenv('NEWS_API_KEY')
+    _news_config = {'client': {'base_url': 'https://newsapi.org/v2/', 'auth': {'type': 'api_key', 'name': 'apiKey', 'api_key': api_key_4, 'location': 'query'}, 'paginator': {'base_page': 1, 'type': 'page_number', 'page_param': 'page', 'total_path': None, 'maximum_page': 3}}, 'resources': [{'name': 'news_articles', 'endpoint': {'path': 'everything', 'params': {'q': 'python', 'language': 'en', 'pageSize': 20}}}]}
+    _news_source = rest_api_source(_news_config)
+    pipeline_4 = dlt.pipeline(pipeline_name='news_pipeline', destination='duckdb', dataset_name='news')
+    pipeline_4.run(_news_source)
+    print(pipeline_4.last_trace)
+    return
+
+
+@app.cell(hide_code=True)
+def _(mo):
+    mo.md(r"""
+    ## **5. Incremental loading**
+
+    Although NewsAPI does not support true incremental loading via cursors, you can simulate it using the `from` or `to` date filters and dlt's `incremental` loader:
+
+    ```python
+    "from": {
+        "type": "incremental",
+        "cursor_path": "publishedAt",
+        "initial_value": "2024-01-01T00:00:00Z",
+    },
+    ```
+
+    This setup means:
+    - dlt will remember the last `publishedAt` seen
+    - On the next run, it will only request articles newer than that
+
+    This is optional and depends on your usage pattern.
+    """)
+    return
+
+
+@app.cell
+def _(dlt, os, rest_api_source):
+    api_key_5 = os.getenv('NEWS_API_KEY')
+    _news_config = {'client': {'base_url': 'https://newsapi.org/v2/', 'auth': {'type': 'api_key', 'name': 'apiKey', 'api_key': api_key_5, 'location': 'query'}, 'paginator': {'base_page': 1, 'type': 'page_number', 'page_param': 'page', 'total_path': None, 'maximum_page': 3}}, 'resources': [{'name': 'news_articles', 'endpoint': {'path': 'everything', 'params': {'q': 'python', 'language': 'en', 'pageSize': 20, 'from': {'type': 'incremental', 'cursor_path': 'publishedAt', 'initial_value': '2025-04-15T00:00:00Z'}}}}]}
```

Comment:
the free plan of `newsapi.org` only allows you to fetch news from a maximum of 1 month ago.
We need to make `initial_value` incremental for this example to work